Factorize CSSValueCurrentColor checks.

mdubet opened the pull request:
#### c69e0a9fdc0ce5ef20afed35a518e44b188b6630
<pre>
Factorize CSSValueCurrentColor checks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245443">https://bugs.webkit.org/show_bug.cgi?id=245443</a>

Reviewed by Tim Nguyen and Chris Dumez.

This is part of the more general refactoring from Color to StyleColor.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::isColorKeyword):
* Source/WebCore/css/StyleColor.h:
(WebCore::StyleColor::isCurrentColorKeyword):
(WebCore::StyleColor::isCurrentColor):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::addMatchedProperties):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::colorAndSVGPaintType):
(WebCore::Style::BuilderCustom::applyValueFill):
(WebCore::Style::BuilderCustom::applyValueStroke):

Canonical link: <a href="https://commits.webkit.org/254731@main">https://commits.webkit.org/254731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/391705a0ff64d399994c7c9badc2430d5efe0d17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90041 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34588 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/99376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/156881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94047 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33080 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82378 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/93672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95689 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/26199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/69199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1417 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -673,7 +673,7 @@ void ElementRuleCollector::addMatchedProperties(MatchedProperties&& matchedPrope
             // The value currentColor has implicitely the same side effect. It depends on the value of color,
             // which is an inherited value, making the non-inherited property implicitly inherited.
-            if (is<CSSPrimitiveValue>(value) && downcast<CSSPrimitiveValue>(value).valueID() == CSSValueCurrentcolor)
+            if (is<CSSPrimitiveValue>(value) && StyleColor::isCurrentColor(downcast<CSSPrimitiveValue>(value)))
```
This might be hot code (not 100% sure but I guess @anttijk would know). If so, it might be unfortunate to now have an extra function call and an extra `isValueID()` check.

cdumez replied:
Maybe we can drop the `isValueID()` check like Sam suggested and inline the function in the header to avoid any perf regression risk?

mdubet replied:
Done 👍

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1443,6 +1443,26 @@ inline void BuilderCustom::applyValueCursor(BuilderState& builderState, CSSValue
     }
 }
 
+inline std::pair<Color, SVGPaintType> helperApplySVGColorValue(BuilderState& builderState, const CSSPrimitiveValue& localValue, String& url)
```
Can we rename this function to something like `colorAndSVGPaintType`?

mdubet replied:
Fixed.

weinig commented on Source/WebCore/css/StyleColor.cpp:
```diff
@@ -77,8 +78,21 @@ bool StyleColor::isSystemColorKeyword(CSSValueID id)
+
+bool StyleColor::isCurrentColor(const CSSPrimitiveValue& value)
+{
+    if (value.isValueID())
```
No need to check isValueID() here. valueID() returns CSSValueInvalid for the case when it is not.

nt1m commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,6 +31,8 @@
 
 #pragma once
 
+
```
nit: unnecessary blank line

nt1m commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -47,6 +49,11 @@ enum class StyleColorOptions : uint8_t {
     static Color colorFromKeyword(CSSValueID, OptionSet<StyleColorOptions>);
     static bool isAbsoluteColorKeyword(CSSValueID);
+
+    // Those functions are inline for performance reason.
```
```suggestion
    // Those functions are inline to avoid any potential performance impact.
```

nt1m replied:
Let's just remove the comment as Chris said.

nt1m commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -47,6 +49,11 @@ enum class StyleColorOptions : uint8_t {
     static bool isAbsoluteColorKeyword(CSSValueID);
+
+    // Those functions are inline for performance reason.
+    ALWAYS_INLINE static bool isCurrentColorKeyword(CSSValueID id) { return (id == CSSValueCurrentcolor); }
```
```suggestion
    ALWAYS_INLINE static bool isCurrentColorKeyword(CSSValueID id) { return id == CSSValueCurrentcolor; }
```

cdumez replied:
We probably don't need the ALWAYS_INLINE here. We should trust the compiler to make the right decision.

Also we do not need the parentheses around the comparison.

cdumez commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -47,6 +49,11 @@ enum class StyleColorOptions : uint8_t {
     static Color colorFromKeyword(CSSValueID, OptionSet<StyleColorOptions>);
     static bool isAbsoluteColorKeyword(CSSValueID);
+
+    // Those functions are inline for performance reason.
```
I don't think this comment is valuable.

cdumez reviewed (approved):
LGTM with nits.

mdubet committed:
Factorize CSSValueCurrentColor checks.
https://bugs.webkit.org/show_bug.cgi?id=245443

Reviewed by Tim Nguyen and Chris Dumez.

This is part of the more general refactoring from Color to StyleColor.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::isColorKeyword):
* Source/WebCore/css/StyleColor.h:
(WebCore::StyleColor::isCurrentColorKeyword):
(WebCore::StyleColor::isCurrentColor):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::addMatchedProperties):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::colorAndSVGPaintType):
(WebCore::Style::BuilderCustom::applyValueFill):
(WebCore::Style::BuilderCustom::applyValueStroke):

Canonical link: https://commits.webkit.org/254731@main
